### PR TITLE
fix: n-plus-1 on dashboard get

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

While working on #9331 I discovered that getting a dashboard has n-plus-1 query performance

The largest part of the problem was:

When loading the dashboard the "last_accessed_at" field was saved with a new value

Also, there's a signal that saved every item on a dashboard whenever the dashboard was changed **in case the filters on the dashboard had changed**

## Changes

removes that save in the case where we know the only change is the last_accessed_at field. 

Also, removes filtering on a foreign ley lookup that is already being filtered earlier in the execution path

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

Adds a test to prove the query count is stable

### CODE REVIEW

File: posthog/api/dashboard.py [hunk 1]
```diff
@@ -161,7 +161,7 @@ def get_items(self, dashboard: Dashboard):
         if self.context["view"].action == "list":
             return None
 
-        items = dashboard.items.filter(deleted=False).order_by("order").all()
```
Comment (Twixes) [inline posthog/api/dashboard.py:164]: This appears to be correct, but I'm not actually sure where the filtering _does_ place. Could you pinpoint?

**Review by Twixes — COMMENTED** [review]
Makes sense 🙌 Left two small comments
Aaand a couple of PR checks are 🔴

File: posthog/api/test/test_dashboard.py [hunk 2]
```diff
@@ -191,6 +193,67 @@ def test_return_cached_results(self):
...
+        )
+        count, qs = self._get_dashboard_counting_queries(dashboard)
+        query_counts.append(count)
+        queries.append(qs)
+
+        # regression test
+        # getting a dashboard was originally n plus 1,
+        # with number of queries growing as the number of insights on the dashboard grew
+        # now a stable 8 queries no matter how many insights are on the dashboard
+        self.assertEqual(query_counts, [8, 8, 8, 8, 8], f"received: {query_counts} for queries: \n\n {queries}")
```
Comment (hazzadous) [inline posthog/api/test/test_dashboard.py:255]: Can we just assert all the counts are the same? That way we won't be tripped up if e.g. a middleware introduces another query.

File: posthog/api/test/test_dashboard.py [hunk 3]
```diff
@@ -191,6 +193,67 @@ def test_return_cached_results(self):
...
+
+        query_counts: List[int] = []
+        queries: List[List[Dict[str, str]]] = []
+
+        count, qs = self._get_dashboard_counting_queries(dashboard)
+        query_counts.append(count)
+        queries.append(qs)
+
+        # add insights to the dashboard and count how manh queries to read the dashboard afterwards
+        Insight.objects.create(
```
Comment (hazzadous) [inline posthog/api/test/test_dashboard.py:223]: Is there an http api for this?

File: posthog/api/dashboard.py (hunk 1 above)
Comment (pauldambra) [reply posthog/api/dashboard.py:164]: It's in `get_queryset` https://github.com/PostHog/posthog/blob/139bcae5b64a4465daac4a42c91d368ed1f4c6dd/posthog/api/dashboard.py#L223

File: posthog/api/test/test_dashboard.py (hunk 3 above)
Comment (pauldambra) [reply posthog/api/test/test_dashboard.py:223]: I'm making that change in #9331  :)

File: posthog/api/test/test_dashboard.py (hunk 2 above)
Comment (pauldambra) [reply posthog/api/test/test_dashboard.py:255]: Hmmm, good point. The output in the error message is where it's useful to know what is happening

**Comment by pauldambra** [comment]
ping @Twixes @hazzadous

File: posthog/api/test/test_dashboard.py [hunk 4]
```diff
@@ -182,15 +184,67 @@ def test_return_cached_results(self):
...
+        filter = Filter(data=filter_dict)
+
+        query_counts: List[int] = []
+        queries: List[List[Dict[str, str]]] = []
+
+        count, qs = self._get_dashboard_counting_queries(dashboard_id)
+        query_counts.append(count)
+        queries.append(qs)
+
+        # add insights to the dashboard and count how manh queries to read the dashboard afterwards
```
Comment (hazzadous) [inline posthog/api/test/test_dashboard.py:219]: What's the significance of adding four insights? If we are checking for O(1) queries then we only need to check two datapoints?

Comment (pauldambra) [reply posthog/api/test/test_dashboard.py:219]: It helped me see how quickly it climbed...

e.g. difference between 2, 4, 6, 8, and 2, 20, 38, 56

File: posthog/api/dashboard.py (hunk 1 above)
Comment (Twixes) [reply posthog/api/dashboard.py:164]: Hmm, I see. But does it hurt to leave the `.filter(deleted=False)` in this line? I'm just a bit afraid someone in the future will use the serializer in a place that doesn't have that in `.prefetch_related()`, resulting in soft-deleted insights showing up.

**Review by hazzadous — APPROVED** [review]

File: posthog/api/dashboard.py (hunk 1 above)
Comment (pauldambra) [reply posthog/api/dashboard.py:164]: 🤔  reasonable concern....

It only adds 1 query, so now reduced from 13 to 9. I can live with that :)

**Pull request merged**